networkx/algorithms/tests/test_graph_hashing.py

```diff
@@ -1,42 +1,624 @@
+        """
+        G = nx.Graph()
+        G.add_nodes_from([(1, {"node_attr1": "a"}), (2, {})])
+        G.add_edges_from([(1, 2), (2, 3), (3, 1), (1, 4)])
+        pytest.raises(
+            KeyError, nx.weisfeiler_lehman_graph_hash, G, node_attr="node_attr1"
+        )
+
+    def test_isomorphic_edge_attr_and_node_attr(self):
+        """
+        Isomorphic graphs with differing node attributes should yield different graph
+        hashes if the 'node_attr' and 'edge_attr' argument is supplied and populated in
+        the graph, and there are no hash collisions.
+        The output should still be invariant to node-relabeling
+        """
+        n, r = 100, 10
+        p = 1.0 / r
+        for i in range(1, r + 1):
+            G1 = nx.erdos_renyi_graph(n, p * i, seed=500 + i)
+
+            for u in G1.nodes():
+                G1.nodes[u]["node_attr1"] = f"{u}-1"
+                G1.nodes[u]["node_attr2"] = f"{u}-2"
+
+            for a, b in G1.edges:
+                G1[a][b]["edge_attr1"] = f"{a}-{b}-1"
+                G1[a][b]["edge_attr2"] = f"{a}-{b}-2"
+
+            g1_hash_edge1_node1 = nx.weisfeiler_lehman_graph_hash(
+                G1, edge_attr="edge_attr1", node_attr="node_attr1"
+            )
+            g1_hash_edge2_node2 = nx.weisfeiler_lehman_graph_hash(
+                G1, edge_attr="edge_attr2", node_attr="node_attr2"
+            )
+            g1_hash_edge1_node2 = nx.weisfeiler_lehman_graph_hash(
+                G1, edge_attr="edge_attr1", node_attr="node_attr2"
+            )
+            g1_hash_no_attr = nx.weisfeiler_lehman_graph_hash(G1)
+
+            assert g1_hash_edge1_node1 != g1_hash_no_attr
+            assert g1_hash_edge2_node2 != g1_hash_no_attr
+            assert g1_hash_edge1_node1 != g1_hash_edge2_node2
+            assert g1_hash_edge1_node2 != g1_hash_edge2_node2
+            assert g1_hash_edge1_node2 != g1_hash_edge1_node1
+
+            G2 = _relabel_graph(G1)
+
+            g2_hash_edge1_node1 = nx.weisfeiler_lehman_graph_hash(
+                G2, edge_attr="edge_attr1", node_attr="node_attr1"
+            )
+            g2_hash_edge2_node2 = nx.weisfeiler_lehman_graph_hash(
+                G2, edge_attr="edge_attr2", node_attr="node_attr2"
+            )
+
+            assert g1_hash_edge1_node1 == g2_hash_edge1_node1
+            assert g1_hash_edge2_node2 == g2_hash_edge2_node2
+
+    def test_digest_size(self):
+        """
+        The hash string lengths should be as expected for a variety of graphs and
+        digest sizes
+        """
+        n, r = 100, 10
+        p = 1.0 / r
+        for i in range(1, r + 1):
+            G = nx.erdos_renyi_graph(n, p * i, seed=1000 + i)
+
+            h16 = nx.weisfeiler_lehman_graph_hash(G)
+            h32 = nx.weisfeiler_lehman_graph_hash(G, digest_size=32)
+
+            assert h16 != h32
+            assert len(h16) == 16 * 2
+            assert len(h32) == 32 * 2
+
+
+class TestSubgraphHashes:
```

Comment:
This class is similarly not used as a class and should be changed to a set of test functions.